### PR TITLE
Server query improvement

### PR DESCRIPTION
**Pull request by paulswartz**
This adds a new type of query to `State.Server`, where we can build more efficient queries by using more features of ETS/Mnesia. There's a lot of code unfortunately, but each commit is self-contained.

### CODE REVIEW

File: apps/state/lib/state/server/query.ex [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        selectors =
+          for record <- records_from_struct_and_values(struct, index, index_values) do
+            {
+              record,
+              guards,
+              [:"$_"]
+            }
+          end
+
+        %{result | selectors: [selectors | result.selectors]}
```
Comment (handorff) [inline apps/state/lib/state/server/query.ex]: Should this be `[selectors ++ result.selectors]`?

File: apps/state/lib/state/server/query.ex [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+  alias __MODULE__.Result
+
+  defp do_query(module, qs) do
+    qs
+    |> Enum.reduce(%Result{module: module}, &accumulate/2)
+    |> finalize()
+  end
+
+  def accumulate(q, %Result{module: module} = result) when map_size(q) > 0 do
```
Comment (handorff) [inline apps/state/lib/state/server/query.ex]: Is this function returning the union of queries when it should be returning the intersection?
e.g. if `q = %{a: [1, 2], b: [3, 4]}` we want to find matches where `(a in [1, 2]) AND (b in [3, 4])`. It seems like we're always appending each individual result to the list of matches or selectors.

Comment (paulswartz) [reply apps/state/lib/state/server/query.ex]: It's the union of intersections, so this function handles one of the intersections, and adds it to the growing union. For example, in State.Stop:
```elixir
  def location_type_0_ids_by_parent_ids(ids) do
    [
      %{
        id: ids,
        location_type: [0]
      },
      %{
        parent_station: ids,
        location_type: [0]
      }
    ]
    |> State.Stop.Cache.query()
    |> Enum.map(& &1.id)
  end
```

This returns `(id IN ids AND location_type = 0) OR (parent_station IN ids AND location_type = 0)`.

File: apps/state/lib/state/server/query.ex [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      {true, :a}
+
+      iex> first_index([:a, :b], %{b: 2})
+      {true, :b}
+
+      iex> first_index([:a, :b], %{c: 3})
+      {false, :c}
+  """
+  @spec first_index([index, ...], q) :: {boolean, index}
+  def first_index(indices, q) do
```
[the remark this answers is not in the document]
Comment (paulswartz) [reply apps/state/lib/state/server/query.ex]: They should already be ordered: earlier we'd gone through and tried to put the more useful indicies earlier in the list, and the primary key (or closest) at the front.

File: apps/state/lib/state/server/query.ex (hunk 1 above)
Comment (paulswartz) [reply apps/state/lib/state/server/query.ex]: No, they get flattened in `finalize`:

```elixir
  def finalize(%Result{module: module, selectors: selectors, matches: matches}) do
    selected = Server.select_with_selectors(module, List.flatten(selectors))
    flat_results([selected | matches])
  end
```

**Comment by paulswartz** [comment]
This doesn't improve the performance in practice, and queries aren't really our bottleneck at the moment. #111 is still useful, so I'll clean that up.

**Pull request closed without merging**